src/Support/ClaimProbabilities.php

```diff
@@ -0,0 +1,179 @@
+<?php
+
+namespace Enjin\Platform\Beam\Support;
+
+use Enjin\Platform\Beam\Enums\PlatformBeamCache;
+use Enjin\Platform\Beam\Models\Beam;
+use Enjin\Platform\Beam\Models\BeamClaim;
+use Enjin\Platform\Beam\Rules\Traits\IntegerRange;
+use Illuminate\Database\Eloquent\Model;
+use Illuminate\Support\Arr;
+use Illuminate\Support\Facades\Cache;
+use Illuminate\Support\Facades\Log;
+
+class ClaimProbabilities
+{
+    use IntegerRange;
+
+    /**
+     * Create or update the probabilities for a claim.
+     */
+    public function createOrUpdateProbabilities(string $code, array $claims): void
+    {
+        $fts = $this->filterClaims($claims);
+        $nfts = $this->filterClaims($claims, true);
+        $current = $this->getProbabilities($code);
+        $this->computeProbabilities(
+            $code,
+            $this->mergeTokens(Arr::get($current, 'tokens.ft', []), $fts),
+            $this->mergeTokens(Arr::get($current, 'tokens.nft', []), $nfts),
+        );
+    }
+
+    /**
+     * Check if the key has probabilities.
+     */
+    public function hasProbabilities(string $code): bool
+    {
+        return Cache::has(PlatformBeamCache::CLAIM_PROBABILITIES->key($code));
+    }
+
+    /**
+     * Get the probabilities for a code.
+     */
+    public function getProbabilities(string $code): array
+    {
+        return Cache::get(PlatformBeamCache::CLAIM_PROBABILITIES->key($code), []);
+    }
+
+    /**
+     * Remove tokens from the probabilities.
+     */
+    public function removeTokens(string $code, array $tokenIds): void
+    {
+        if ($current = $this->getProbabilities($code)) {
+            foreach (Arr::get($current, 'tokens.ft', []) as $tokenId => $quantity) {
+                if (in_array($tokenId, $tokenIds)) {
+                    unset($current['tokens']['ft'][$tokenId]);
+                }
+            }
+
+            foreach (Arr::get($current, 'tokens.nft', []) as $tokenId => $quantity) {
+                if (in_array($tokenId, $tokenIds)) {
+                    unset($current['tokens']['nft'][$tokenId]);
+                }
+            }
+
+            $this->computeProbabilities(
+                $code,
+                Arr::get($current, 'tokens.ft'),
+                Arr::get($current, 'tokens.nft'),
+            );
+        }
+    }
+
+    /**
+     * Draw a claim base from the probabilities.
+     */
+    public function drawClaim(string $code, ?int $randomInt = null): ?Model
```

Comment:
I'm not sure we need this, the probabilities are meant just for display purposes only and should be returned in the beam type, we don't need to use them on the platform to select a token, the current selection method is fine 🙂.  We don't need to update the probabilities after each claim either, we would only need to update the probabilities if the number of tokens in a beam changes due to adding or removing them.